### PR TITLE
[SPARK-15927] Eliminate redundant DAGScheduler code.

### PR DESCRIPTION
**Pull request by kayousterhout**
## What changes were proposed in this pull request?

To try to eliminate redundant code to traverse the RDD dependency graph, this PR creates a new function getShuffleDependencies that returns shuffle dependencies that are immediate parents of a given RDD.  This new function is used by getParentStages and getAncestorShuffleDependencies.

cc @squito @markhamstra  

FYI @rxin

### CODE REVIEW

**Comment by JoshRosen** [comment]
I think that @cloud-fan proposed a similar change in #4134, so I think we might be able to resolve SPARK-5374 as a duplicate of this JIRA.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -412,25 +390,48 @@ class DAGScheduler(
...
     waitingForVisit.push(rdd)
     while (waitingForVisit.nonEmpty) {
-      visit(waitingForVisit.pop())
+      val toVisit = waitingForVisit.pop()
+      if (!visited(toVisit)) {
+        visited += toVisit
+        toVisit.dependencies.foreach {
+          case shuffleDep: ShuffleDependency[_, _, _] =>
+            parents += shuffleDep
+          case dependency: Any =>
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: nit: we can remove `: Any`

**Comment by cloud-fan** [comment]
LGTM

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 2]
```diff
@@ -412,25 +390,48 @@ class DAGScheduler(
...
+        visited += toVisit
+        getShuffleDependencies(toVisit).foreach { shuffleDep =>
+          if (!shuffleToMapStage.contains(shuffleDep.shuffleId)) {
+            parents.push(shuffleDep)
+            waitingForVisit.push(shuffleDep.rdd)
+          } // Otherwise, the dependency and it's ancestors have already been registered.
         }
       }
     }
+    parents
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: while you're touching this, would be nice to rename `parents` to `ancestors` here.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 3]
```diff
@@ -378,31 +378,9 @@ class DAGScheduler(
...
-    val parents = new HashSet[Stage]
-    val visited = new HashSet[RDD[_]]
-    // We are manually maintaining a stack here to prevent StackOverflowError
-    // caused by recursively visiting
-    val waitingForVisit = new Stack[RDD[_]]
-    def visit(r: RDD[_]) {
-      if (!visited(r)) {
-        visited += r
-        // Kind of ugly: need to register RDDs with the cache here since
-        // we can't do it in its constructor because # of partitions is unknown
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: this comment is getting lost ... but I think that is OK, I have no idea what it is referring to here.  Its also in `newOrUsedShuffleStage`, which is probably the only place it belongs.

**Comment by squito** [comment]
lgtm.  I realized there isn't really any test for going through a mix of narrow and shuffle dependencies -- might be nice to add this (it passes w/ this change).  You would need to change the visibility of `getShuffleDependencies` to `private[scheduler]` though:

```scala
  /**
   * Ensure we get all parent shuffle stages.  Walk back all narrow dependencies, but don't go
   * past any shuffle dependencies (parent stages only, not ancestor stages).
   */
  test("getShuffleDependencies") {
    val rddA = new MyRDD(sc, 2, Nil)
    val shuffleDepA = new ShuffleDependency(rddA, new HashPartitioner(1))
    val rddB = new MyRDD(sc, 2, Nil)
    val shuffleDepB = new ShuffleDependency(rddB, new HashPartitioner(1))
    val rddC = new MyRDD(sc, 1, List(shuffleDepB))
    val shuffleDepC = new ShuffleDependency(rddC, new HashPartitioner(1))
    val rddD = new MyRDD(sc, 1, List(shuffleDepC))
    val narrowDepD = new OneToOneDependency(rddD)
    val finalRdd = new MyRDD(sc, 1, List(shuffleDepA, narrowDepD), tracker = mapOutputTracker)

    assert(scheduler.getShuffleDependencies(rddA) === Set())
    assert(scheduler.getShuffleDependencies(rddB) === Set())
    assert(scheduler.getShuffleDependencies(rddC) === Set(shuffleDepB))
    assert(scheduler.getShuffleDependencies(rddD) === Set(shuffleDepC))
    assert(scheduler.getShuffleDependencies(finalRdd) === Set(shuffleDepA, shuffleDepC))
  }
```

**Comment by markhamstra** [comment]
LGTM, but I agree with Imran's renaming suggestion, and his new test looks good.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 3 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Yeah I decided that it was probably an accident that it's here -- as you said, I think it only belongs in newOrUsedShuffleStage.

**Comment by kayousterhout** [comment]
Thanks for the comments / test @squito!  Updated the naming and the test. 

Once this is merged, I have another DAGScheduler PR that does some more renaming / commenting / re-organizing to try to make some of the code easier to read / understand.

**Comment by kayousterhout** [comment]
Jenkins, retest this please

**Comment by kayousterhout** [comment]
(@squito this looks like another timeout in the basic scheduler integration suite? I tried it locally and it passed)

**Comment by squito** [comment]
lgtm!

not sure about the test timeout.  I will take a look, but in any case I'm confident its not from this change.  I've tried running that test in a loop but haven't triggered a failure in over 10k trials so far :(

**Comment by kayousterhout** [comment]
Merged into master (not 2.0).

**Comment by squito** [comment]
@kayousterhout btw I tracked down the test issue.  It wasn't safe to check the map output status inside the backend in those tests -- I addressed that here: https://github.com/apache/spark/pull/13565/commits/91ea3df33aaebb20b2df0cbbe56dbc99975200ea.  Also the real problem was somewhat hidden b/c the assertion was failing in another thread, which I fixed here: https://github.com/apache/spark/pull/13565/commits/e35a7d3d0696fd0c864ee8fa9c9ded95abffba2c

I ran the tests over 15k times on my laptop and still never triggered it, though.  wish I had a better way to find these issues.

**Comment by JoshRosen** [comment]
It looks like a bunch of the tests in `BasicSchedulerIntegrationSuite` are now really flaky. For example: https://spark-tests.appspot.com/tests/org.apache.spark.scheduler.BasicSchedulerIntegrationSuite/multi-stage%20job

Is there any way that we can hotfix / merge in the small fix for this sooner rather than later, perhaps by separating it from the blacklist test PR into its own PR?

**Comment by squito** [comment]
yikes, thanks for pointing this out @JoshRosen .  Sorry I was wrong about the failure being unrelated -- it didn't create a new bug, but took an old rare problem and made it far more common.  I have opened https://github.com/apache/spark/pull/13688.

**Comment by kayousterhout** [comment]
Ah shoot I see why this would make the stage ID numbering potentially different.  Sorry about that!!

**Pull request closed without merging**